File: platform/android/VoiceChatRollbackAnchor.kt
package com.voicechat.crypto.storage

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.io.File
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Rollback anchor for dycrpt on Android.
 *
 * ## Read this before shipping
 *
 * Android exposes **no** app-accessible hardware monotonic counter. There is no
 * API that gives a value which provably only increases and cannot be reverted
 * by an attacker with the device. Anyone claiming otherwise is describing
 * `setRollbackResistant`, which is about key *deletion*, not counters.
 *
 * So be precise about what this class does and does not buy you.
 *
 * ### What it defends against
 *
 * **Backup/restore rollback.** The counter is sealed with a non-exportable
 * Android Keystore key and written to `noBackupFilesDir`. Neither the Keystore
 * key nor the file participates in Android Backup, Auto Backup, or a
 * device-to-device transfer. Restoring an old app data set therefore yields a
 * counter that cannot be decrypted (Keystore key absent or different), which
 * surfaces to Rust as `VC_STATE_LOST` — terminal, fail-closed, correct.
 *
 * This is the common real-world case: a user restores a backup, or moves to a
 * new phone, and the stale ratchet state must not be reused.
 *
 * ### What it does NOT defend against
 *
 * **An attacker with root or a live-device exploit.** Such an attacker can copy
 * `counter.bin` and the state file together at time T and put both back at time
 * T+n. The Keystore key is still present, the counter still decrypts, and the
 * pair is internally consistent — so the rollback is undetectable by this
 * anchor. Defending against that requires an anchor the attacker does not
 * control: a **server-held counter**.
 *
 * If your threat model includes a compromised device, use
 * [ServerRollbackAnchor] instead and treat this class as unsuitable.
 *
 * ### Failure mode you must handle
 *
 * If the user removes their lock screen, the Keystore key can be invalidated
 * and the counter becomes permanently undecryptable. The engine will report
 * `VC_STATE_LOST`. That is fail-closed and intended — but it is also a real
 * user-facing lockout, and your app must show something better than a crash.
 *
 * Thread-safety: every method is `@Synchronized`. The Rust engine calls these
 * from multiple threads, and the JNI bridge does not add locking of its own.
 */
class VoiceChatRollbackAnchor(private val context: Context) {

    companion object {
        private const val KEYSTORE = "AndroidKeyStore"
        private const val ALIAS = "dycrpt-anchor-seal-v1"
        private const val TRANSFORMATION = "AES/GCM/NoPadding"
        private const val VERSION: Byte = 1
        private const val IV_BYTES = 12
    }

    private val counterFile: File
        get() = File(context.noBackupFilesDir, "dycrpt-anchor.v1")

    /** Current committed anchor value. Returns 0 for a never-provisioned device. */
    @Synchronized
    fun current(): Long {
        val file = counterFile
        if (!file.exists()) return 0L
        return decodeSealed(file.readBytes())
    }

    /**
     * Atomically move [expected] to `expected + 1` and return the new value.
     *
     * Returns -1 if the current value is not [expected]; the Rust side treats a
     * negative return as failure. The durable value is not modified on any
     * failure path: the write is temp-file + fsync + atomic rename, so the
     * outcome is never left unknown, which is what the anchor contract demands.
     */
    @Synchronized
    fun compareAndIncrement(expected: Long): Long {
        if (expected < 0) return -1
        val observed = current()
        if (observed != expected) return -1
        val next = expected + 1
        atomicWrite(counterFile, encodeSealed(next))
        return next
    }

    /** Destroy the anchor. Only for deliberate, user-visible device reset. */
    @Synchronized
    fun destroy() {
        counterFile.delete()
        val ks = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        if (ks.containsAlias(ALIAS)) ks.deleteEntry(ALIAS)
    }

    private fun sealingKey(): SecretKey {
        val ks = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        (ks.getKey(ALIAS, null) as? SecretKey)?.let { return it }
        val gen = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE)
        gen.init(
            KeyGenParameterSpec.Builder(
                ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setKeySize(256)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setRandomizedEncryptionRequired(true)
                .build()
        )
        return gen.generateKey()
    }

    private fun encodeSealed(value: Long): ByteArray {
        val plain = ByteArray(8)
        for (i in 0 until 8) plain[i] = ((value shr (8 * i)) and 0xff).toByte()
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, sealingKey())
        val iv = cipher.iv
        require(iv.size == IV_BYTES)
        val ct = cipher.doFinal(plain)
        return ByteArray(1 + IV_BYTES + ct.size).also {
            it[0] = VERSION
            System.arraycopy(iv, 0, it, 1, IV_BYTES)
            System.arraycopy(ct, 0, it, 1 + IV_BYTES, ct.size)
        }
    }

    private fun decodeSealed(blob: ByteArray): Long {
        require(blob.size >= 1 + IV_BYTES + 16) { "corrupt dycrpt anchor" }
        require(blob[0] == VERSION) { "unsupported dycrpt anchor version" }
        val iv = blob.copyOfRange(1, 1 + IV_BYTES)
        val ct = blob.copyOfRange(1 + IV_BYTES, blob.size)
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, sealingKey(), GCMParameterSpec(128, iv))
        val plain = cipher.doFinal(ct)
        require(plain.size == 8) { "corrupt dycrpt anchor payload" }
        var v = 0L
        for (i in 7 downTo 0) v = (v shl 8) or (plain[i].toLong() and 0xff)
        return v
    }

    private fun atomicWrite(target: File, bytes: ByteArray) {
        target.parentFile?.mkdirs()
        val temp = File(target.parentFile, ".${target.name}.${System.nanoTime()}.tmp")
        temp.outputStream().use { s ->
            s.write(bytes)
            s.flush()
            s.fd.sync()
        }
        if (!temp.renameTo(target)) {
            temp.delete()
            throw IllegalStateException("unable to persist dycrpt anchor")
        }
    }
}

/**
 * Server-anchored counter — the option that actually survives a compromised
 * device, because the attacker does not hold the counter.
 *
 * This is an interface, not an implementation, because the correctness lives in
 * your backend, not here. The endpoint must:
 *
 * 1. store the counter per (account, device) and increment **atomically** —
 *    a single conditional-update statement, not read-then-write;
 * 2. reject an increment whose `expected` does not match the stored value;
 * 3. never return success unless the write is durably committed;
 * 4. on network failure, leave the client able to **re-read** and learn the
 *    true value. An increment whose outcome is unknown is exactly the state the
 *    anchor contract forbids, so the client must resolve it by re-reading
 *    before returning, and must surface an error rather than guess.
 *
 * Point 4 is the one that is usually got wrong. A timeout after the server
 * committed, treated as failure, desynchronizes the epoch and is
 * indistinguishable from a rollback at the next open.
 */
interface ServerRollbackAnchor {
    /** Committed value; 0 if this device has never been provisioned. */
    fun current(): Long

    /** Atomic CAS from [expected] to `expected + 1`; -1 if it did not apply. */
    fun compareAndIncrement(expected: Long): Long
}

File: platform/ios/VoiceChatRollbackAnchor.swift
import Foundation
import Security

/// Rollback anchor for dycrpt on iOS.
///
/// ## Read this before shipping
///
/// iOS exposes no app-accessible hardware monotonic counter. The Secure Enclave
/// signs and wraps; it does not vend a counter that provably only increases.
///
/// ### What this defends against
///
/// **Backup/restore rollback.** The counter is a Keychain generic-password item
/// with `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`. Items with a
/// `ThisDeviceOnly` protection class are excluded from iCloud backup and from
/// encrypted iTunes/Finder backups, and do not migrate to a new device.
///
/// So restoring an old app container, or moving to a new phone, leaves the
/// counter absent while the state file returns. That surfaces to Rust as
/// `VC_STATE_LOST` — terminal, fail-closed, correct.
///
/// ### What this does NOT defend against
///
/// **A jailbroken or otherwise compromised device.** An attacker with Keychain
/// access can capture the counter and state file together and replay both. The
/// pair stays internally consistent and the rollback is invisible here.
/// Defending against that requires an anchor the attacker does not hold — a
/// server-side counter. See `ServerRollbackAnchor`.
///
/// ### Ordering requirement
///
/// `commit` writes the counter with `kSecAttrAccessibleAfterFirstUnlock…`, so it
/// is unavailable before first unlock after boot. If your app performs crypto
/// from a background push before first unlock, `current()` will fail and the
/// engine will report `VC_ANCHOR_UNAVAILABLE`. That is a transient, retryable
/// condition and must NOT be treated as a rollback — the Rust layer already
/// gives you a distinct code for exactly this reason.
///
/// Thread-safety: guarded by a serial queue. The Rust engine is concurrent and
/// the C shim adds no locking.
public final class VoiceChatRollbackAnchor {

    public enum AnchorError: Error {
        case keychain(OSStatus)
        case corrupt
    }

    private let service = "com.voicechat.crypto.dycrpt.anchor"
    private let account = "dycrpt-anchor-v1"
    private let queue = DispatchQueue(label: "com.voicechat.crypto.dycrpt.anchor")

    public init() {}

    /// Committed anchor value; 0 for a never-provisioned device.
    public func current() throws -> UInt64 {
        try queue.sync {
            guard let data = try loadRaw() else { return 0 }
            guard data.count == 8 else { throw AnchorError.corrupt }
            return data.withUnsafeBytes { $0.loadUnaligned(as: UInt64.self).littleEndian }
        }
    }

    /// Atomic move from `expected` to `expected + 1`. Returns the new value.
    /// Throws if the stored value is not `expected`; the durable value is never
    /// modified on a failure path.
    public func compareAndIncrement(expected: UInt64) throws -> UInt64 {
        try queue.sync {
            let observed: UInt64
            if let data = try loadRaw() {
                guard data.count == 8 else { throw AnchorError.corrupt }
                observed = data.withUnsafeBytes { $0.loadUnaligned(as: UInt64.self).littleEndian }
            } else {
                observed = 0
            }
            guard observed == expected else { throw AnchorError.corrupt }

            let next = expected + 1
            var le = next.littleEndian
            let payload = Data(bytes: &le, count: 8)
            try store(payload, replacingExisting: observed != 0 || (try loadRaw()) != nil)
            return next
        }
    }

    /// Destroy the anchor. Only for deliberate, user-visible device reset.
    public func destroy() throws {
        try queue.sync {
            let q: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrService as String: service,
                kSecAttrAccount as String: account,
            ]
            let status = SecItemDelete(q as CFDictionary)
            guard status == errSecSuccess || status == errSecItemNotFound else {
                throw AnchorError.keychain(status)
            }
        }
    }

    private func loadRaw() throws -> Data? {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var out: CFTypeRef?
        let status = SecItemCopyMatching(q as CFDictionary, &out)
        switch status {
        case errSecSuccess:
            return out as? Data
        case errSecItemNotFound:
            return nil
        default:
            throw AnchorError.keychain(status)
        }
    }

    private func store(_ data: Data, replacingExisting: Bool) throws {
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        if replacingExisting {
            let status = SecItemUpdate(
                base as CFDictionary,
                [kSecValueData as String: data] as CFDictionary
            )
            if status == errSecSuccess { return }
            if status != errSecItemNotFound { throw AnchorError.keychain(status) }
        }
        var add = base
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(add as CFDictionary, nil)
        guard status == errSecSuccess else { throw AnchorError.keychain(status) }
    }
}

// MARK: - C bridge

/// Boxed context handed to the Rust engine as `VcRollbackAnchorCallbacks.ctx`.
///
/// Must outlive the engine handle. `VoiceChatPersistentEngine` owns it.
public final class AnchorBox {
    let anchor: VoiceChatRollbackAnchor
    init(_ a: VoiceChatRollbackAnchor) { self.anchor = a }
}

private func anchorCurrent(_ ctx: UnsafeMutableRawPointer?, _ out: UnsafeMutablePointer<UInt64>?)
    -> Int32
{
    guard let ctx, let out else { return 1 }
    let box = Unmanaged<AnchorBox>.fromOpaque(ctx).takeUnretainedValue()
    do {
        out.pointee = try box.anchor.current()
        return 0
    } catch {
        return 1
    }
}

private func anchorCas(
    _ ctx: UnsafeMutableRawPointer?, _ expected: UInt64, _ out: UnsafeMutablePointer<UInt64>?
) -> Int32 {
    guard let ctx, let out else { return 1 }
    let box = Unmanaged<AnchorBox>.fromOpaque(ctx).takeUnretainedValue()
    do {
        out.pointee = try box.anchor.compareAndIncrement(expected: expected)
        return 0
    } catch {
        return 1
    }
}

/// Opens the persistent Rust engine with Keychain-backed storage key and anchor.
///
/// The returned object owns the anchor box; releasing it while the engine handle
/// is alive would leave Rust holding a dangling `ctx`, which for a *rollback*
/// anchor means silently losing rollback detection rather than crashing.
public final class VoiceChatPersistentEngine {

    public enum OpenError: Error {
        /// Terminal. Stale state, or state lost. Do not retry, do not
        /// re-provision silently — see the header docs on recovery policy.
        case rollbackDetected
        case stateLost
        /// Transient. Anchor unreadable, e.g. before first unlock.
        case anchorUnavailable
        /// Device was never provisioned; call `provision` instead.
        case notInitialized
        case native(Int32)
    }

    public let handle: VcHandle
    private let box: AnchorBox

    public static func open(
        deviceId: Data, profile: UInt8, path: String, storageKey: Data,
        anchor: VoiceChatRollbackAnchor, createIfAbsent: Bool
    ) throws -> VoiceChatPersistentEngine {
        precondition(storageKey.count == 32, "storage key must be 32 bytes")
        let box = AnchorBox(anchor)
        let cbs = VcRollbackAnchorCallbacks(
            ctx: Unmanaged.passUnretained(box).toOpaque(),
            current: anchorCurrent,
            compare_and_increment: anchorCas
        )

        var handle: VcHandle = 0
        let pathBytes = Array(path.utf8)
        let rc: Int32 = deviceId.withUnsafeBytes { dev in
            storageKey.withUnsafeBytes { key in
                vc_engine_open_persistent(
                    dev.bindMemory(to: UInt8.self).baseAddress, deviceId.count,
                    profile,
                    pathBytes, pathBytes.count,
                    key.bindMemory(to: UInt8.self).baseAddress,
                    cbs,
                    createIfAbsent ? 1 : 0,
                    &handle,
                    nil
                )
            }
        }

        switch rc {
        case 0: return VoiceChatPersistentEngine(handle: handle, box: box)
        case 7: throw OpenError.rollbackDetected
        case 8: throw OpenError.stateLost
        case 9: throw OpenError.anchorUnavailable
        case 10: throw OpenError.notInitialized
        default: throw OpenError.native(rc)
        }
    }

    private init(handle: VcHandle, box: AnchorBox) {
        self.handle = handle
        self.box = box
    }

    deinit {
        _ = vc_engine_destroy(handle)
        // `box` is released only after the engine is destroyed, so no in-flight
        // native call can still reach through the raw ctx pointer.
    }
}
